lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -963,6 +963,11 @@ OpFoldResult MuxPrimOp::fold(ArrayRef<Attribute> operands) {
   if (getType().getBitWidthOrSentinel() < 0)
     return {};
 
+  // // mux(0-bit, x, y) -> y
+  if (sel().getType().cast<FIRRTLType>().getBitWidthOrSentinel() == 0 &&
+      low().getType() == getType())
+    return low();
+
```

Comment:
I think this might need an additional check to ensure that the type of the mux and its arguments have a known width. The mux has an implicit `pad` operation that must be preserved across folds. For example:
```mlir
%c2_ui = firrtl.constant 2 : !firrtl.uint  // infers to uint<2>
%c4_ui = firrtl.constant 4 : !firrtl.uint  // infers to uint<3>
%0 = firrtl.mux(%sel, %c4_ui, %c2_ui) : (!firrtl.uint<0>, !firrtl.uint, !firrtl.uint) -> !firrtl.uint
%z = firrtl.node %0 : !firrtl.uint
```
Without the zero-width canonicalization, this would infer to:
```mlir
%c2_ui2 = firrtl.constant 2 : !firrtl.uint<2>
%c4_ui3 = firrtl.constant 4 : !firrtl.uint<3>
%0 = firrtl.mux(%sel, %c4_ui3, %c2_ui2) :
  (!firrtl.uint<0>, !firrtl.uint<3>, !firrtl.uint<2>) -> !firrtl.uint<3>
%z = firrtl.node %0 : !firrtl.uint<3>
```
With the canonicalization, if the mux picks one of the operands before width inference (where it cannot know which size to pad to):
```mlir
%c2_ui = firrtl.constant 2 : !firrtl.uint
%z = firrtl.node %c2_ui : !firrtl.uint
// eventually infers to
%c2_ui2 = firrtl.constant 2 : !firrtl.uint<2>
%z = firrtl.node %c2_ui2 : !firrtl.uint<2>  // <-- wrong type
```
If the mux picks one of the operands after width inference:
```mlir
%c2_ui2 = firrtl.constant 2 : !firrtl.uint<2>
%0 = firrtl.pad(%c2_ui2) : (!firrtl.uint<2>) -> !firrtl.uint<3>
%z = firrtl.node %0 : !firrtl.uint<3>
```
I think there might be some other parts of mux canonicalization that also needs to handle this case already :thinking:

Reply:
Yeah I think you're right about this.  Adding `hasKnownWidthIntTypesAndNonZeroResult` as a condition is good enough I think.

Reply:
Digging around here and I think the example code is unreachable.  The check just before this (https://github.com/llvm/circt/pull/1349/commits/84f3a9f280f8355a610fb7fcd7f3d60273d4cea4#diff-cf1544e8c0ef055dbe799b8b3f07e3e3a443422f7452981466c81be44619515aR963) that the mux type (the return type) has a known width. 

The implementation of the `InferTypeOpInterface` (via the `inferReturnType` method) on `MuxPrimOp` is going to infer the return type of the mux as having unknown width if either the true or false paths have unknown width.

Therefore, I don't think this canonicalization will ever run if either the true or false paths have uninferred widths.

Running a slightly modified version of the above example (`%z` is now an output port) doesn't cause any changes to the circuit for `circt-opt -simple-canonicalizer`:

```mlir
module  {
  firrtl.circuit "Foo" {
    firrtl.module @Foo (in %sel: !firrtl.uint<0>, out %z: !firrtl.uint) {
      %c2_ui = firrtl.constant 2 : !firrtl.uint
      %c4_ui = firrtl.constant 4 : !firrtl.uint
      %0 = firrtl.mux(%sel, %c2_ui, %c4_ui) :
        (!firrtl.uint<0>, !firrtl.uint, !firrtl.uint) -> !firrtl.uint
      firrtl.connect %z, %0 : !firrtl.uint, !firrtl.uint
    }
  }
}
```